runtime/stdlib/random.go

```diff
@@ -195,3 +205,207 @@ func NewRevertibleRandomFunction(generator RandomGenerator) StandardLibraryValue
+		var moduloVal interpreter.UInt64Value
+		moduloVal, ok = moduloArg.(interpreter.UInt64Value)
+		modulo = uint64(moduloVal)
+	case sema.Word8Type:
+		var moduloVal interpreter.Word8Value
+		moduloVal, ok = moduloArg.(interpreter.Word8Value)
+		modulo = uint64(moduloVal)
+	case sema.Word16Type:
+		var moduloVal interpreter.Word16Value
+		moduloVal, ok = moduloArg.(interpreter.Word16Value)
+		modulo = uint64(moduloVal)
+	case sema.Word32Type:
+		var moduloVal interpreter.Word32Value
+		moduloVal, ok = moduloArg.(interpreter.Word32Value)
+		modulo = uint64(moduloVal)
+	case sema.Word64Type:
+		var moduloVal interpreter.Word64Value
+		moduloVal, ok = moduloArg.(interpreter.Word64Value)
+		modulo = uint64(moduloVal)
+	default:
+		// sanity check: shouldn't reach here
+		panic(errors.NewUnreachableError())
+	}
+
+	if !ok {
+		// checker should prevent this
+		panic(errors.NewUnreachableError())
+	}
+
+	// user error if modulo is zero
+	if modulo == 0 {
+		panic(ZeroModuloError)
+	}
+
+	// `max` is the maximum value that can be returned
+	max := modulo - 1
+	// get a bit mask (0b11..11) that covers all `max` bits,
+	// and count the byte size of `max`
+	mask := uint64(0)
+	bitSize := 0
+	for max&mask != max {
+		bitSize++
+		mask = (mask << 1) | 1
+	}
+	byteSize := (bitSize + 7) >> 3
+
+	// Generate a number less or equal than `max`.
+	// use the reject-sample method to avoid the modulo bias.
+	// the function isn't constant-time in this case and may take longer than computing
+	// a modular reduction.
+	// However, sampling exactly the size of `max` in bits makes the loop return fast:
+	// loop returns after (k) iterations with a probability of at most 1-(1/2)^k.
+	//
+	// (a different approach would be to pull 128 bits more bits than the size of `max`
+	// from the random generator and use big number reduction by `modulo`)
+	random := modulo
+	for random > max {
+		// only generate `byteSize` random bytes
+		getRandomBytes(buffer[bufferSize-byteSize:], generator)
+		// big endianness must be used in this case
+		random = binary.BigEndian.Uint64(buffer[:])
+		// adjust to the size of max in bits
+		random &= mask
+	}
+	return random
+}
+
+// cases of a random number of size larger than 8 bytes can be all treated
+// by the same function, based on the big.Int type.
+func getBigRandomNumber(
+	generator RandomGenerator,
+	ty sema.Type,
+	moduloArg interpreter.Value,
+) *big.Int {
+
+	// buffer to get random bytes from the generator
+	// 32 is the size of the largest type supported
+	const bufferSize = 32
+	var buffer [bufferSize]byte
+	// case where no modulo argument was provided
+	if moduloArg == nil {
+		numericType, ok := ty.(*sema.NumericType)
+		if !ok {
+			// checker should prevent this
+			panic(errors.NewUnreachableError())
+		}
+		bytes := numericType.ByteSize()
+		getRandomBytes(buffer[:bytes], generator)
+		// SetBytes considers big endianness (although little endian could be used too)
+		return new(big.Int).SetBytes(buffer[:bytes])
```

Comment:
following my comment above:
```suggestion
		buffer := make([]byte, bytes)
		getRandomBytes(buffer, generator)
		// SetBytes considers big endianness (although little endian could be used too)
		return new(big.Int).SetBytes(buffer)
```